File: #0416.partition-equal-subset-sum.cpp
class Solution {
public:
    bool canPartition(vector<int>& nums) {

        std::bitset<10001> bits(1);
        int sum = std::accumulate(nums.begin(), nums.end(), 0);
        for (auto num : nums) {
            bits |= bits << num;
        }
        return !(sum & 1) && bits[sum >> 1];
    }
};

File: #0416.partition-equal-subset-sum.js
/**
 * @param {number[]} nums
 * @return {boolean}
 */
var canPartition = function(nums) {

    const total = nums.reduce((a, b) => a + b, 0);
    if (total % 2) {
        return false;
    }

    const target = total / 2;
    let everySum = new Set([0]);
    for (let i = nums.length - 1; i >= 0; i--) {
        const nextSums = new Set()
        for (const sum of everySum) {
            if (nums[i] + sum == target) {
                return true;
            }
            nextSums.add(sum);
            nextSums.add(sum + nums[i]);
        }
        everySum = nextSums;
    }
    
    return everySum.has(target);
};
